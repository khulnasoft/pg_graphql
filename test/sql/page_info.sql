begin;
    create table account(
        id int primary key,
        is_verified bool
    );


    -- hasNextPage and hasPreviousPage should be non-null on empty collection
    -- startCursor and endCursor may be null
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
          }
        }
        $$)
    );

    insert into account(id) select generate_series(1, 10);

    -- Forward pagination
    -- hasPreviousPage is false, hasNextPage is true
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(first: 5) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );

    -- hasPreviousPage is true, hasNextPage is true
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(first: 5, after: "WzJd" ) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );

    -- hasPreviousPage is true, hasNextPage is false
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(first: 5, after: "Wzdd" ) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );

    -- Backward pagination
    -- hasPreviousPage is true, hasNextPage is false
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(last: 5) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );

    -- hasPreviousPage is true, hasNextPage is true
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(last: 5, before: "Wzdd" ) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );

    -- hasPreviousPage is false, hasNextPage is true
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(last: 5, before: "WzJd" ) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );

    -- OFFSET ARG
    -- Only supported for forward pagination


    -- hasPreviousPage is false, hasNextPage is true
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(first: 2, offset: 0) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );

    -- hasPreviousPage is true, hasNextPage is true
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(first: 2, offset: 1) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );

    -- hasPreviousPage is true, hasNextPage is true
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(
            first: 1,
            offset: 1,
            after: "Wzdd" # id = 7
        ) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );

    -- hasPreviousPage is true, hasNextPage is false
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(
            first: 2,
            offset: 2,
            after: "Wzdd" # id = 7
        ) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );


    -- Error States

    -- Offset doesn't work with "last"
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(
            last: 2,
            offset: 2,
        ) {
            edges {
              node {
                id
              }
            }
          }
        }
        $$)
    );

    -- Offset doesn't work with "before"
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(
            last: 2,
            offset: 2,
        ) {
            edges {
              node {
                id
              }
            }
          }
        }
        $$)
    );

rollback;
