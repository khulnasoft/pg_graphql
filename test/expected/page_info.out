begin;
    create table account(
        id int primary key,
        is_verified bool
    );
    -- hasNextPage and hasPreviousPage should be non-null on empty collection
    -- startCursor and endCursor may be null
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
          }
        }
        $$)
    );
               jsonb_pretty               
------------------------------------------
 {                                       +
     "data": {                           +
         "accountCollection": {          +
             "pageInfo": {               +
                 "endCursor": null,      +
                 "hasNextPage": false,   +
                 "startCursor": null,    +
                 "hasPreviousPage": false+
             }                           +
         }                               +
     }                                   +
 }
(1 row)

    insert into account(id) select generate_series(1, 10);
    -- Forward pagination
    -- hasPreviousPage is false, hasNextPage is true
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(first: 5) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );
               jsonb_pretty               
------------------------------------------
 {                                       +
     "data": {                           +
         "accountCollection": {          +
             "edges": [                  +
                 {                       +
                     "node": {           +
                         "id": 1         +
                     },                  +
                     "cursor": "WzFd"    +
                 },                      +
                 {                       +
                     "node": {           +
                         "id": 2         +
                     },                  +
                     "cursor": "WzJd"    +
                 },                      +
                 {                       +
                     "node": {           +
                         "id": 3         +
                     },                  +
                     "cursor": "WzNd"    +
                 },                      +
                 {                       +
                     "node": {           +
                         "id": 4         +
                     },                  +
                     "cursor": "WzRd"    +
                 },                      +
                 {                       +
                     "node": {           +
                         "id": 5         +
                     },                  +
                     "cursor": "WzVd"    +
                 }                       +
             ],                          +
             "pageInfo": {               +
                 "endCursor": "WzVd",    +
                 "hasNextPage": true,    +
                 "startCursor": "WzFd",  +
                 "hasPreviousPage": false+
             }                           +
         }                               +
     }                                   +
 }
(1 row)

    -- hasPreviousPage is true, hasNextPage is true
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(first: 5, after: "WzJd" ) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );
              jsonb_pretty               
-----------------------------------------
 {                                      +
     "data": {                          +
         "accountCollection": {         +
             "edges": [                 +
                 {                      +
                     "node": {          +
                         "id": 3        +
                     },                 +
                     "cursor": "WzNd"   +
                 },                     +
                 {                      +
                     "node": {          +
                         "id": 4        +
                     },                 +
                     "cursor": "WzRd"   +
                 },                     +
                 {                      +
                     "node": {          +
                         "id": 5        +
                     },                 +
                     "cursor": "WzVd"   +
                 },                     +
                 {                      +
                     "node": {          +
                         "id": 6        +
                     },                 +
                     "cursor": "WzZd"   +
                 },                     +
                 {                      +
                     "node": {          +
                         "id": 7        +
                     },                 +
                     "cursor": "Wzdd"   +
                 }                      +
             ],                         +
             "pageInfo": {              +
                 "endCursor": "Wzdd",   +
                 "hasNextPage": true,   +
                 "startCursor": "WzNd", +
                 "hasPreviousPage": true+
             }                          +
         }                              +
     }                                  +
 }
(1 row)

    -- hasPreviousPage is true, hasNextPage is false
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(first: 5, after: "Wzdd" ) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );
               jsonb_pretty               
------------------------------------------
 {                                       +
     "data": {                           +
         "accountCollection": {          +
             "edges": [                  +
                 {                       +
                     "node": {           +
                         "id": 8         +
                     },                  +
                     "cursor": "Wzhd"    +
                 },                      +
                 {                       +
                     "node": {           +
                         "id": 9         +
                     },                  +
                     "cursor": "Wzld"    +
                 },                      +
                 {                       +
                     "node": {           +
                         "id": 10        +
                     },                  +
                     "cursor": "WzEwXQ=="+
                 }                       +
             ],                          +
             "pageInfo": {               +
                 "endCursor": "WzEwXQ==",+
                 "hasNextPage": false,   +
                 "startCursor": "Wzhd",  +
                 "hasPreviousPage": true +
             }                           +
         }                               +
     }                                   +
 }
(1 row)

    -- Backward pagination
    -- hasPreviousPage is true, hasNextPage is false
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(last: 5) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );
               jsonb_pretty               
------------------------------------------
 {                                       +
     "data": {                           +
         "accountCollection": {          +
             "edges": [                  +
                 {                       +
                     "node": {           +
                         "id": 6         +
                     },                  +
                     "cursor": "WzZd"    +
                 },                      +
                 {                       +
                     "node": {           +
                         "id": 7         +
                     },                  +
                     "cursor": "Wzdd"    +
                 },                      +
                 {                       +
                     "node": {           +
                         "id": 8         +
                     },                  +
                     "cursor": "Wzhd"    +
                 },                      +
                 {                       +
                     "node": {           +
                         "id": 9         +
                     },                  +
                     "cursor": "Wzld"    +
                 },                      +
                 {                       +
                     "node": {           +
                         "id": 10        +
                     },                  +
                     "cursor": "WzEwXQ=="+
                 }                       +
             ],                          +
             "pageInfo": {               +
                 "endCursor": "WzEwXQ==",+
                 "hasNextPage": false,   +
                 "startCursor": "WzZd",  +
                 "hasPreviousPage": true +
             }                           +
         }                               +
     }                                   +
 }
(1 row)

    -- hasPreviousPage is true, hasNextPage is true
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(last: 5, before: "Wzdd" ) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );
              jsonb_pretty               
-----------------------------------------
 {                                      +
     "data": {                          +
         "accountCollection": {         +
             "edges": [                 +
                 {                      +
                     "node": {          +
                         "id": 2        +
                     },                 +
                     "cursor": "WzJd"   +
                 },                     +
                 {                      +
                     "node": {          +
                         "id": 3        +
                     },                 +
                     "cursor": "WzNd"   +
                 },                     +
                 {                      +
                     "node": {          +
                         "id": 4        +
                     },                 +
                     "cursor": "WzRd"   +
                 },                     +
                 {                      +
                     "node": {          +
                         "id": 5        +
                     },                 +
                     "cursor": "WzVd"   +
                 },                     +
                 {                      +
                     "node": {          +
                         "id": 6        +
                     },                 +
                     "cursor": "WzZd"   +
                 }                      +
             ],                         +
             "pageInfo": {              +
                 "endCursor": "WzZd",   +
                 "hasNextPage": true,   +
                 "startCursor": "WzJd", +
                 "hasPreviousPage": true+
             }                          +
         }                              +
     }                                  +
 }
(1 row)

    -- hasPreviousPage is false, hasNextPage is true
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(last: 5, before: "WzJd" ) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );
               jsonb_pretty               
------------------------------------------
 {                                       +
     "data": {                           +
         "accountCollection": {          +
             "edges": [                  +
                 {                       +
                     "node": {           +
                         "id": 1         +
                     },                  +
                     "cursor": "WzFd"    +
                 }                       +
             ],                          +
             "pageInfo": {               +
                 "endCursor": "WzFd",    +
                 "hasNextPage": true,    +
                 "startCursor": "WzFd",  +
                 "hasPreviousPage": false+
             }                           +
         }                               +
     }                                   +
 }
(1 row)

    -- OFFSET ARG
    -- Only supported for forward pagination
    -- hasPreviousPage is false, hasNextPage is true
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(first: 2, offset: 0) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );
               jsonb_pretty               
------------------------------------------
 {                                       +
     "data": {                           +
         "accountCollection": {          +
             "edges": [                  +
                 {                       +
                     "node": {           +
                         "id": 1         +
                     },                  +
                     "cursor": "WzFd"    +
                 },                      +
                 {                       +
                     "node": {           +
                         "id": 2         +
                     },                  +
                     "cursor": "WzJd"    +
                 }                       +
             ],                          +
             "pageInfo": {               +
                 "endCursor": "WzJd",    +
                 "hasNextPage": true,    +
                 "startCursor": "WzFd",  +
                 "hasPreviousPage": false+
             }                           +
         }                               +
     }                                   +
 }
(1 row)

    -- hasPreviousPage is true, hasNextPage is true
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(first: 2, offset: 1) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );
              jsonb_pretty               
-----------------------------------------
 {                                      +
     "data": {                          +
         "accountCollection": {         +
             "edges": [                 +
                 {                      +
                     "node": {          +
                         "id": 2        +
                     },                 +
                     "cursor": "WzJd"   +
                 },                     +
                 {                      +
                     "node": {          +
                         "id": 3        +
                     },                 +
                     "cursor": "WzNd"   +
                 }                      +
             ],                         +
             "pageInfo": {              +
                 "endCursor": "WzNd",   +
                 "hasNextPage": true,   +
                 "startCursor": "WzJd", +
                 "hasPreviousPage": true+
             }                          +
         }                              +
     }                                  +
 }
(1 row)

    -- hasPreviousPage is true, hasNextPage is true
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(
            first: 1,
            offset: 1,
            after: "Wzdd" # id = 7
        ) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );
              jsonb_pretty               
-----------------------------------------
 {                                      +
     "data": {                          +
         "accountCollection": {         +
             "edges": [                 +
                 {                      +
                     "node": {          +
                         "id": 9        +
                     },                 +
                     "cursor": "Wzld"   +
                 }                      +
             ],                         +
             "pageInfo": {              +
                 "endCursor": "Wzld",   +
                 "hasNextPage": true,   +
                 "startCursor": "Wzld", +
                 "hasPreviousPage": true+
             }                          +
         }                              +
     }                                  +
 }
(1 row)

    -- hasPreviousPage is true, hasNextPage is false
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(
            first: 2,
            offset: 2,
            after: "Wzdd" # id = 7
        ) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
        $$)
    );
                jsonb_pretty                
--------------------------------------------
 {                                         +
     "data": {                             +
         "accountCollection": {            +
             "edges": [                    +
                 {                         +
                     "node": {             +
                         "id": 10          +
                     },                    +
                     "cursor": "WzEwXQ=="  +
                 }                         +
             ],                            +
             "pageInfo": {                 +
                 "endCursor": "WzEwXQ==",  +
                 "hasNextPage": false,     +
                 "startCursor": "WzEwXQ==",+
                 "hasPreviousPage": true   +
             }                             +
         }                                 +
     }                                     +
 }
(1 row)

    -- Error States
    -- Offset doesn't work with "last"
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(
            last: 2,
            offset: 2,
        ) {
            edges {
              node {
                id
              }
            }
          }
        }
        $$)
    );
                                   jsonb_pretty                                    
-----------------------------------------------------------------------------------
 {                                                                                +
     "data": null,                                                                +
     "errors": [                                                                  +
         {                                                                        +
             "message": "\"offset\" may only be used with \"first\" and \"after\""+
         }                                                                        +
     ]                                                                            +
 }
(1 row)

    -- Offset doesn't work with "before"
    select jsonb_pretty(
        graphql.resolve($$
        {
          accountCollection(
            last: 2,
            offset: 2,
        ) {
            edges {
              node {
                id
              }
            }
          }
        }
        $$)
    );
                                   jsonb_pretty                                    
-----------------------------------------------------------------------------------
 {                                                                                +
     "data": null,                                                                +
     "errors": [                                                                  +
         {                                                                        +
             "message": "\"offset\" may only be used with \"first\" and \"after\""+
         }                                                                        +
     ]                                                                            +
 }
(1 row)

rollback;
